File: tools/mkinitrd.c
/*
 * mkinitrd.c -- Build an initrd image from a list of files.
 *
 * Usage: mkinitrd <output> <file1> [file2] ...
 *
 * Format: [count:4] { [name:64] [size:4] [data:size] } ...
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>

#define NAME_LEN 64

int main(int argc, char **argv)
{
    FILE *out;
    uint32_t count;
    int i;

    if (argc < 3) {
        fprintf(stderr, "Usage: %s <output> <file1> [file2] ...\n", argv[0]);
        return 1;
    }

    out = fopen(argv[1], "wb");
    if (!out) { perror(argv[1]); return 1; }

    count = (uint32_t)(argc - 2);
    fwrite(&count, 4, 1, out);

    for (i = 2; i < argc; i++) {
        FILE *in;
        char name[NAME_LEN];
        uint32_t size;
        char *buf;
        const char *basename;

        /* Extract basename */
        basename = strrchr(argv[i], '/');
        basename = basename ? basename + 1 : argv[i];

        memset(name, 0, NAME_LEN);
        strncpy(name, basename, NAME_LEN - 1);

        in = fopen(argv[i], "rb");
        if (!in) { perror(argv[i]); fclose(out); return 1; }

        fseek(in, 0, SEEK_END);
        size = (uint32_t)ftell(in);
        fseek(in, 0, SEEK_SET);

        buf = malloc(size);
        fread(buf, 1, size, in);
        fclose(in);

        fwrite(name, 1, NAME_LEN, out);
        fwrite(&size, 4, 1, out);
        fwrite(buf, 1, size, out);
        free(buf);
    }

    fclose(out);
    printf("Created initrd with %u files\n", count);
    return 0;
}

File: tools/mkinitrd.rs
use std::env;
use std::fs;
use std::io::Write;
use std::path::Path;

const NAME_LEN: usize = 64;

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() < 3 {
        eprintln!("Usage: {} <output> <file1> [file2] ...", args[0]);
        std::process::exit(1);
    }
    let mut out = fs::File::create(&args[1]).expect("cannot create output file");
    let count = (args.len() - 2) as u32;
    out.write_all(&count.to_le_bytes()).unwrap();
    for i in 2..args.len() {
        let path = Path::new(&args[i]);
        let basename = path.file_name().unwrap().to_str().unwrap();
        let mut name = [0u8; NAME_LEN];
        let bytes = basename.as_bytes();
        let n = bytes.len().min(NAME_LEN - 1);
        name[..n].copy_from_slice(&bytes[..n]);
        let data = fs::read(&args[i]).expect("cannot read file");
        let size = data.len() as u32;
        out.write_all(&name).unwrap();
        out.write_all(&size.to_le_bytes()).unwrap();
        out.write_all(&data).unwrap();
    }
    println!("Created initrd with {} files", count);
}
